src/http/client.rs

```diff
@@ -20,14 +20,31 @@ impl Default for HttpClient {
 
 impl HttpClient {
   pub fn new(server: Server) -> Self {
+    let upstream_settings = &server.upstream.clone().unwrap_or(Upstream {
```

Comment:
Please, note with current code defining upstream in config require settings all the fields of it, as if you omit property it won't be properly parsed by serde and the whole `server.upstream` will be None. I.e. there is no way to override specific settings without defining the whole set through config.

To mitigate the issue I suggest to use [serde default](https://serde.rs/attr-default.html) attribute for fields of upstream that will provide defaults properly. 

Another suggestion it's actually write some basic test to catch the behaviour I've mentioned as it's not obvious 🙂 

Another thing I've noticed that in case of failed parsing there is no error on `tc check` and therefore some default logic just applying. It's worth another issue I think

Reply:
@meskill yeah I agree with you, in present case client have to define all the fields. 

Previously I played with [derivative](https://mcarton.github.io/rust-derivative/latest/index.html). Its pretty straightforward. But this project is not using this crate. Maybe we can use this ? not sure.

And regarding tests, a comment from tushar

> I think we can ignore the tests, as this is something that we can rely on for the compiler for. It is not possible to create an invalid client using the builder.

Reply:
I'm not sure that derivative could help in this case as it seems like allows to generate implementation for `Default` trait with some customization, but here we need not just default for the whole struct but defaults for every field separately when deserializing it.

About test, just wanted to point out that compiler verification sometimes is not enough to be sure everything works as expected 🙂

Reply:
I agree with @meskill I think we should add some tests around the serde bits of `Server`. We can add a simple test where we assert that default is empty.

```rust
Server::default().to_json() == "{}"
```

@meskill I think we could remove the option wrapping from `Server` as well.